lib/vsc/utils/cli.py

```diff
@@ -0,0 +1,193 @@
+#
+# This file is part of vsc-utils,
+# originally created by the HPC team of Ghent University (http://ugent.be/hpc/en),
+# with support of Ghent University (http://ugent.be/hpc),
+# the Flemish Supercomputer Centre (VSC) (https://www.vscentrum.be),
+# the Flemish Research Foundation (FWO) (http://www.fwo.be/en)
+# and the Department of Economy, Science and Innovation (EWI) (http://www.ewi-vlaanderen.be/en).
+#
+# https://github.com/hpcugent/vsc-utils
+#
+# vsc-utils is free software: you can redistribute it and/or modify
+# it under the terms of the GNU Library General Public License as
+# published by the Free Software Foundation, either version 2 of
+# the License, or (at your option) any later version.
+#
+# vsc-utils is distributed in the hope that it will be useful,
+# but WITHOUT ANY WARRANTY; without even the implied warranty of
+# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the
+# GNU Library General Public License for more details.
+#
+# You should have received a copy of the GNU Library General Public License
+# along with vsc-utils. If not, see <http://www.gnu.org/licenses/>.
+#
+"""
+Main kafka based class for producers and consumer
+"""
+
+import logging
+import json
+
+from datetime import datetime, timedelta
+
+from kafka import KafkaProducer, KafkaConsumer
+
+from vsc.utils.timestamp import convert_to_datetime
+from vsc.utils.script_tools import NrpeCLI
+from enum import Enum
+
+
+def make_time(ts, fmt="%Y-%m-%d", begin=False, end=False):
+    """
+    Return timestamp in format fmt
+
+    If begin if True, round down till begin of day
+    If end is True, round up till end of day
+    """
+
+    dt = convert_to_datetime(ts)
+    if begin or end:
+        dt = datetime(*dt.timetuple()[:3])
+        if end:
+            # add 1 day, subtract 1 second
+            dt += timedelta(days=1)
+            dt -= timedelta(seconds=1)
+
+    return datetime.strftime(dt, fmt)
+
+
+class Ingestor(Enum):
+    slurm = 'slurm'
+    openstack = 'openstack'
+    cloudresourcespecs = 'cloudresourcespecs'
+
+
+class KafkaCLI(NrpeCLI):
+    """
+    Base class for Kafka based NrpeCLI clients
+    """
+    KAFKA_COMMON_OPTIONS = {
+        'topic': ("Kafka topics to produce/consume", None, "store", "xdmod"),
+        'brokers': ("List of kafka brokers, comma separated", "strlist", "store", None),
+        'security_protocol': ("Security protocol to use, e.g., SASL_SSL", str, "store", "PLAINTEXT"),
+        'ssl': ("Comma-separated key=value list of SSL options for underlying kafka lib", "strlist", "store", []),
+        'sasl': ("Comma-separated key=value list of SASL options for the underlying kafka lib", "strlist", "store", []),
+        # Very advanced/dangerous usage:
+        # e.g. on initial usage of consumer, pass --kafka=auto_offset_reset=earliest
+        #    to start from the earliest offset in abscence of (first) commit
+        'kafka': ("Comma-separated key=value list of allowed options for the underlying kafka lib",
```

Comment:
if this is dangerous, maybe put an extra option so people have to use 2 options, or put the word `dangerous` in the description.

Reply:
I think Stijn added that comment. Normally one should not know things that are dangerous. in any case, we use it all the time to pass the credentials and stuff.